extensions-contrib/opentelemetry-extensions/src/main/java/org/apache/druid/data/input/opentelemetry/protobuf/OpenTelemetryMetricsProtobufReader.java

```diff
@@ -0,0 +1,232 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.data.input.opentelemetry.protobuf;
+
+import com.google.common.collect.Maps;
+import com.google.common.collect.Sets;
+import com.google.protobuf.InvalidProtocolBufferException;
+import io.opentelemetry.proto.metrics.v1.InstrumentationLibraryMetrics;
+import io.opentelemetry.proto.metrics.v1.IntDataPoint;
+import io.opentelemetry.proto.metrics.v1.Metric;
+import io.opentelemetry.proto.metrics.v1.MetricsData;
+import io.opentelemetry.proto.metrics.v1.NumberDataPoint;
+import io.opentelemetry.proto.metrics.v1.ResourceMetrics;
+import org.apache.druid.data.input.InputEntityReader;
+import org.apache.druid.data.input.InputRow;
+import org.apache.druid.data.input.InputRowListPlusRawValues;
+import org.apache.druid.data.input.MapBasedInputRow;
+import org.apache.druid.data.input.impl.ByteEntity;
+import org.apache.druid.data.input.impl.DimensionsSpec;
+import org.apache.druid.java.util.common.CloseableIterators;
+import org.apache.druid.java.util.common.parsers.CloseableIterator;
+import org.apache.druid.java.util.common.parsers.ParseException;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+
+public class OpenTelemetryMetricsProtobufReader implements InputEntityReader
+{
+
+  private final ByteEntity source;
+  private final String metricDimension;
+  private final String valueDimension;
+  private final String metricAttributePrefix;
+  private final String resourceAttributePrefix;
+  private final DimensionsSpec dimensionsSpec;
+
+  public OpenTelemetryMetricsProtobufReader(
+      DimensionsSpec dimensionsSpec,
+      ByteEntity source,
+      String metricDimension,
+      String valueDimension,
+      String metricAttributePrefix,
+      String resourceAttributePrefix
+  )
+  {
+    this.dimensionsSpec = dimensionsSpec;
+    this.source = source;
+    this.metricDimension = metricDimension;
+    this.valueDimension = valueDimension;
+    this.metricAttributePrefix = metricAttributePrefix;
+    this.resourceAttributePrefix = resourceAttributePrefix;
+  }
+
+  @Override
+  public CloseableIterator<InputRow> read()
+  {
+    return CloseableIterators.withEmptyBaggage(readAsList().iterator());
+  }
+
+  List<InputRow> readAsList()
+  {
+    try {
+      return parseMetricsData(MetricsData.parseFrom(source.getBuffer()));
+    }
+    catch (InvalidProtocolBufferException e) {
+      throw new ParseException(e, "Protobuf message could not be parsed");
+    }
+  }
+
+  private List<InputRow> parseMetricsData(final MetricsData metricsData)
+  {
+    List<InputRow> rows = new ArrayList<>(getRowCount(metricsData));
```

Comment:
I wonder if the cost of counting points outweights the cost of resizing the array if necessary. Unless we have some evidence this speeds things up, I would say we don't bother with pre-sizing the array.

Reply:
the approach with pre-sizing is somewhat faster than resizing for the single datapoint per message case. Both methods seem to be faster than the OpenCensus InputFormat:

With ForEach:
<img width="1341" alt="Screen Shot 2022-01-28 at 11 08 05 AM" src="https://user-images.githubusercontent.com/16495842/151606938-9f37967d-c6b5-43f1-a374-25cc2212982a.png">

With Stream:
<img width="1337" alt="Screen Shot 2022-01-28 at 11 09 01 AM" src="https://user-images.githubusercontent.com/16495842/151606960-0b14982f-1c3a-43a7-92af-209974db719e.png">

Reply:
it's hard to tell, it seems to be within the error bounds of both benchmarks

Reply:
we might have to run the benchmark on EC2 to get tighter error-bounds.

we should also understand the impact of this optimization for the more general case where there are more data points per payload. If it makes things much worse in those cases it might not really be worth it to keep it only for a single point.

Reply:
That makes sense. On my local machine the numbers were roughly identical between the two methods for 2,4,8 and 16 datapoints. 

How do you feel about having splitting the further optimization work into another ticket? So the follow on work would be:
- Removing the deprecated methods once the upstream collectors are upgraded.
- Changing the default `metricName` dimension to "metric" for the `OpenCensusMetricsInputFormat` so that it's consistent with the `OpenTelemetryMetricsInputFormat`.
- Evaluating/optimizing the `OpenTelemetryMetricsInputFormat`'s performance for different data shapes and data sizes in a cloud instance